keps/sig-network/4559-kubelet-probes/README.md

```diff
@@ -0,0 +1,1037 @@
+
+#### RESOLVED Summary discussions (preserving for history)
+- Figure out if there are any concerns with non-namespace-based Linux
+  runtimes (kata, kubevirt, etc): There are no concerns here because both kata containers and [kubevirt] use network namespaces. They expect the CNI to set up the pod networking. Then their agent creates the required networking infrastructure inside this pod network namespace to connect to the VM interface
+
+#### UNRESOLVED Summary discussions
+
+- Figure out details of exec probe performance problem.
+
+- Figure out if network probe performance is still a problem.
+
+- Figure out if there are any Windows-specific concerns (e.g, that our
+  proposal may be more difficult on Windows). Alternatively, it may
+  turn out that removing the constraint that kubelet needs to be able
+  to talk to all pods would simplify things on Windows?
+
+[kubernetes #99425]: https://github.com/kubernetes/kubernetes/issues/99425
+[AdminNetworkPolicy]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/2091-admin-network-policy/README.md
+[future successor to NetworkPolicy]: https://github.com/kubernetes-sigs/network-policy-api/issues/136
+[multi-tenant scenarios]: https://github.com/kubernetes/kubernetes/issues/31824
+[Multi-network feature]: https://github.com/kubernetes/enhancements/issues/3698
+[kubernetes #101324]: https://github.com/kubernetes/kubernetes/issues/101324
+[the NetworkPolicy has not yet been fully programmed]: https://kubernetes.io/docs/concepts/services-networking/network-policies/#pod-lifecycle
+[readiness gates]: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate
+[Exec probes are much, much slower than network probes]: https://medium.com/netcracker/exec-probes-a-story-about-experiment-and-relevant-problems-12de616c0a76
+[kubernetes #89898]: https://github.com/kubernetes/kubernetes/issues/89898
+[kubernetes #115143]: https://github.com/kubernetes/kubernetes/pull/115143
+[kubernetes #102613]: https://github.com/kubernetes/kubernetes/issues/102613
+[kubevirt]: https://github.com/kubevirt/kubevirt/blob/main/docs/devel/networking.md#vmi-networking
+
+## Motivation
+
+### Goals
+
+```
+<<[UNRESOLVED goals ]>>
+
+The exact details of the goals depend on whether we decide to
+deprecate the existing probe types, or just change their semantics.
+
+Also, there may be additional goals depending on the resolution of the
+UNRESOLVED section in the Summary.
+
+<<[/UNRESOLVED]>>
+```
+
+- Continue to backward-compatibly support all existing probes by
+  default (at least for a while).
+
+- Deprecate the [`TCPSocketAction.Host`] and [`HTTPGetAction.Host`]
+  fields in pod probes (since they allow the SSRF attack and have no
+  other compelling use case).
+
+    - Allow administrators to use [Pod Security admission] to block
+      Pods that have probes using those fields.
+
+- Move to a system where TCP, HTTP, and GRPC pod probes are run inside
```

Comment:
We don't currently support localhost probes though, right?

> may actually need to specify an IP or host or interface?

In general, except for localhost, the user won't know the IP ahead of time.

I feel like support for Multi-Network may need to be a new extension to the probe data once we've defined how Multi-Network works. (Though if we said you specified by interface then that would solve the localhost case right away...)

This also sort of ties into the "probe the non-default IP family" case (in that you want to probe a non-default IP but don't know ahead of time what that IP is).

Reply:
> We don't currently support localhost probes though, right?

No, but if I knew that probes ran from within a netns, I might wonder why I can't use localhost :)

I am happy to NOT do this, as long as we choose not to do it, rather than forget to consider it.